File: solutions/zig/square-root/2/square_root.zig
pub fn squareRoot(radicand: usize) usize {
    var left: usize = 0;
    var right: usize = radicand + 1;

    while (left < right) {
        const mid = left + (right - left) / 2;
        const square = mid * mid;

        if (square == radicand) return (mid)
        else if (square < radicand) left = mid + 1
        else right = mid;
    }
    return 0;
}

File: solutions/elixir/square-root/1/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand), do: calculate(radicand, 0, div(radicand, 2) + 1)

  defp calculate(_ , left, right) when left >= right, do: right
  defp calculate(radicand, left, right) do
    mid = div(left + right, 2)
    square = mid * mid
    cond do
      square == radicand -> mid
      square < radicand -> calculate(radicand, mid + 1, right)
      square > radicand -> calculate(radicand, left, mid)
    end
  end
end
